### DISCUSSION TITLE
Detect if a draft for a new entry is being created in save hook

### DISCUSSION THREAD

**Discussion by MoritzLost** [craftcms/cms#10370] [Ideas] [answered]
I'm trying to write a hook that sets a default value for a newly created entry. Surprisingly, detecting if an element is completely new, not saved and published yet is surprisingly difficult. I'm looking at [this SE post](https://craftcms.stackexchange.com/a/38046/11705) by @brandonkelly for reference. Looks like `firstSave` is only true once the draft is converted to a published entry, so that is too late. There's also the `ElementEvent::isNew` property, though I couldn't find clear documentation on when this is used exactly. Based on [this issue](https://github.com/craftcms/cms/issues/4589) it looks like it also fires once the entry is published.

Based on the other properties, the closest I can get is `$element->isDraft && $element->isCanonical`, so a draft that is canonical  (not a draft of a published entry). But that will also be true when an unpublished draft is edited. Is there no way to detect that a draft is `completely` new in a way that will **only** trigger when a new draft for a new entry is created?

As a broader note, would be great to have this information in the [event code generator](https://craftcms.com/docs/3.x/extend/events.html#event-code-generator) in the docs. Right now there's only an _Element is new_ checkbox, but it's ambiguous what 'new' means in this context.

**Comment by brandonkelly** [comment, marked as answer]
I’m not sure what you’re looking for actually exists… Craft doesn’t let you create a new entry _and_ a new draft for it at the exact same time. Entries will always start in one of these states:

- As a full-fledged entry, when created via the Quick Post widget or an entry editor slideout (which you can identify via `$event->isNew && $event->sender->firstSave`).
- As an unpublished draft, when created via the “New entry” button on the Entries index page (which you’ve indeed found out how to identify – or you could check `$event->isNew && $event->sender->getIsUnpublishedDraft()`).

In neither case will an additional draft be created right off the bat, though.

**Reply by MoritzLost** [reply]
@brandonkelly Thanks Brandon! The second bullet point hits the nail – basically I want to detect if an unpublished draft is being created to fill out a field with a default value. But I absolutely need to be sure that this hook will only trigger if the unpublished draft is completely fresh, so my hook won't overwrite existing values at any point. So `$event->isNew && $event->sender->getIsUnpublishedDraft()` will only be true if it's a newly created unpublished draft, right?

With hooks like this I'm just worried that there are scenarios where my hook triggers unintentionally. For example, cloning an entry, applying an unpublished draft, restoring a revision … and with all the different ways that 'new' entries are defined (for example, `$event->isNew` vs `$element->firstSave`) it's not always easy which of those is used for which scenarios. Maybe this can be simplified a bit in Craft 4? Or maybe some of those need a bit more documentation – for example, the [docblock for `isNew`](https://docs.craftcms.com/api/v3/craft-events-elementevent.html#public-properties) is kind of self-referential, no use without a definition of 'new' in this context. Maybe it's just inherent complexity, not sure.

Anyway, thanks for the pointers!

**Reply by brandonkelly** [reply]
> So `$event->isNew && $event->sender->getIsUnpublishedDraft()` will only be true if it's a newly created unpublished draft, right?

Correct. Entries will only ever be an unpublished draft at the start of their existence, and that unpublished draft will only be _new_ when it is first created.

> For example, cloning an entry

Currently, “Save as a new entry” will currently duplicate it as-is, however the plan is to change that in Craft 4, so entries get duplicated to unpublished drafts (#10244). If you want to ignore that scenario, add `&& !$event->sender->duplicateOf` to your condition:

```php
if ($event->isNew && $event->sender->getIsUnpublishedDraft() && !$event->sender->duplicateOf) {
    // ...
}
```

> Maybe this can be simplified a bit in Craft 4? Or maybe some of those need a bit more documentation – for example, the [docblock for `isNew`](https://docs.craftcms.com/api/v3/craft-events-elementevent.html#public-properties) is kind of self-referential, no use without a definition of 'new' in this context. Maybe it's just inherent complexity, not sure.

Not sure there’s much room to simplify. Ultimately there’s just several states an entry can be in (unpublished draft, normal draft, revision, current entry), and multiple scenarios for what may be happening to the entry. We just launched that [event code generator](https://craftcms.com/docs/3.x/extend/events.html#event-code-generator) recently, and will continue to improve it (as well as the class reference and other documentation).

**Reply by MoritzLost** [reply]
@brandonkelly Thank you for the detailed explanation!
